File: .agents/skills/frappe-app-development/assets/mini-app-template/your_app/report/sample_report/sample_report.py
import frappe
from frappe import _
from frappe.query_builder import functions as fn


def execute(filters=None):
	filters = filters or {}

	columns = [
		{
			"label": _("Title"),
			"fieldname": "title",
			"fieldtype": "Data",
			"width": 200,
		},
		{
			"label": _("Status"),
			"fieldname": "status",
			"fieldtype": "Data",
			"width": 100,
		},
	]

	SampleDoc = frappe.qb.DocType("Sample Doc")
	base = frappe.qb.from_(SampleDoc)

	where_clauses = []
	if filters.get("status"):
		where_clauses.append(SampleDoc.status == filters["status"])
	if filters.get("title"):
		where_clauses.append(SampleDoc.title.like(f"%{filters['title']}%"))

	query = base.select(SampleDoc.name, SampleDoc.title, SampleDoc.status).orderby(
		SampleDoc.modified, order=frappe.qb.desc
	)
	for clause in where_clauses:
		query = query.where(clause)

	page_length = int(filters.get("page_length") or 20)
	page = int(filters.get("page") or 1)
	if page < 1:
		page = 1
	if page_length < 1:
		page_length = 20

	offset = (page - 1) * page_length
	query = query.limit(page_length).offset(offset)

	data = query.run(as_dict=True)

	count_query = base.select(fn.Count("*").as_("count"))
	for clause in where_clauses:
		count_query = count_query.where(clause)
	total_count = count_query.run(as_dict=True)[0]["count"]

	report_summary = [
		{
			"label": _("Total Records"),
			"value": total_count,
			"indicator": "blue",
		}
	]

	return columns, data, None, None, report_summary

File: .agents/skills/frappe-app-development/assets/mini-app-template/your_app/report/sample_report/sample_report.js
frappe.query_reports["Sample Report"] = {
	filters: [
		{
			fieldname: "status",
			label: "Status",
			fieldtype: "Select",
			options: "\nOpen\nClosed",
			default: "Open",
		},
		{
			fieldname: "title",
			label: "Title Contains",
			fieldtype: "Data",
		},
		{
			fieldname: "page_length",
			label: "Page Size",
			fieldtype: "Int",
			default: 20,
		},
		{
			fieldname: "page",
			label: "Page",
			fieldtype: "Int",
			default: 1,
		},
	],
	onload(report) {
		const updatePagerState = () => {
			const current = report.get_filter_value("page") || 1;
			const pageLength = report.get_filter_value("page_length") || 20;
			const summary = report.get_summary ? report.get_summary() : [];
			const totalRow = (summary || []).find((r) => r.label === "Total Records");
			const total = totalRow ? totalRow.value : null;
			const totalPages = total ? Math.max(1, Math.ceil(total / pageLength)) : null;

			if (report.page && report.page.set_title_sub && totalPages) {
				report.page.set_title_sub(`Page ${current} of ${totalPages}`);
			}

			const prevBtn = report.page && report.page.prev_btn;
			const nextBtn = report.page && report.page.next_btn;
			if (prevBtn) {
				prevBtn.prop("disabled", current <= 1);
			}
			if (nextBtn && totalPages) {
				nextBtn.prop("disabled", current >= totalPages);
			}
		};

		const addPagerButtons = () => {
			report.page.prev_btn = report.page.add_inner_button("Prev", () => {
				const current = report.get_filter_value("page") || 1;
				report.set_filter_value("page", Math.max(1, current - 1));
				report.refresh();
			});
			report.page.next_btn = report.page.add_inner_button("Next", () => {
				const current = report.get_filter_value("page") || 1;
				report.set_filter_value("page", current + 1);
				report.refresh();
			});
		};

		addPagerButtons();

		const originalRefresh = report.refresh.bind(report);
		report.refresh = () => {
			const result = originalRefresh();
			if (result && result.then) {
				return result.then(() => updatePagerState());
			}
			setTimeout(updatePagerState, 0);
			return result;
		};
	},
};
